lib/Dialect/TensorExt/Transforms/SIMDPack.cpp

```diff
@@ -0,0 +1,189 @@
+                    "than or equal to ")
+              << numSlots_ << " , got " << type;
+          return std::nullopt;
+        }
+        auto padding = DenseI64ArrayAttr::get(
+            ctx,
+            llvm::ArrayRef<int64_t>(llvm::PowerOf2Ceil(dimension) - dimension));
+
+        return RankedTensorType::get(
+            newShape, type.getElementType(),
+            SIMDPackingAttr::get(
+                ctx, DenseI64ArrayAttr::get(ctx, type.getShape()), padding,
+                DenseI64ArrayAttr::get(ctx, llvm::ArrayRef<int64_t>(newShape)),
+                0));
+      }
+      return type;
+    });
+
+    addConversion([&](secret::SecretType secretType) -> Type {
+      auto convertedTensorType = this->convertType(secretType.getValueType());
+      return secret::SecretType::get(convertedTensorType);
+    });
+
+    numSlots_ = numSlots;
+  }
+  int numSlots_;
+};
+
+struct ConvertTensorEncodingGenericPattern : public ConversionPattern {
+  using ConversionPattern::ConversionPattern;
+
+  ConvertTensorEncodingGenericPattern(SecretTensorTypeConverter &converter,
+                                      MLIRContext *context)
+      : ConversionPattern(converter, RewritePattern::MatchAnyOpTypeTag(),
+                          /*benefit=*/1, context) {}
+
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    SmallVector<Type> newResultTypes;
+    if (failed(getTypeConverter()->convertTypes(op->getResultTypes(),
+                                                newResultTypes)))
+      return failure();
+
+    mlir::OperationState newOpState(op->getLoc(), op->getName().getStringRef(),
+                                    operands, newResultTypes, op->getAttrs(),
+                                    op->getSuccessors());
+
+    for (mlir::Region &region : op->getRegions()) {
+      mlir::Region *newRegion = newOpState.addRegion();
+      rewriter.inlineRegionBefore(region, *newRegion, newRegion->begin());
+
+      mlir::TypeConverter::SignatureConversion signatureConv(
+          newRegion->getNumArguments());
+      if (failed(typeConverter->convertSignatureArgs(
+              newRegion->getArgumentTypes(), signatureConv)))
+        return mlir::failure();
+      rewriter.applySignatureConversion(&newRegion->front(), signatureConv);
+    }
+
+    rewriter.replaceOp(op, rewriter.create(newOpState));
+    return success();
+  }
+};
+
+struct ConvertGenericOp : public OpConversionPattern<secret::GenericOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  ConvertGenericOp(MLIRContext *context)
+      : OpConversionPattern<secret::GenericOp>(context, 2) {}
+
+  LogicalResult matchAndRewrite(
+      secret::GenericOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<mlir::Type> newResultTypes;
+    if (failed(typeConverter->convertTypes(op->getResultTypes(),
```

Comment:
This is a dumb question, but why do you need to convert result types? Shouldn't that already be done on the OpAdaptor? I'm not sure why there's a pattern needed for `secret.generic` at all...

Reply:
It's also not clear to me why we need the other pattern above.

Reply:
I will tinker around with this pass locally to think about it.

Reply:
I don't think the default pattern works for a generic since you have to update the block arguments as well - the op adaptor doesn't expose type converted results, so you have to manually convert result types.

Reply:
Hmm actually the generic one also handles regions so it probably should

Reply:
This just feels like boilerplate. There's no logic specific to the actual transformation happening...

Reply:
I think Utils.cpp https://github.com/google/heir/blob/fc09e806d75ac228c8875f72cb565ceb5e165286/lib/Conversion/Utils.cpp#L30

has a variant of this, if I expose it in the header i'll see if it works as a drop in

Reply:
FYI: https://discord.com/channels/636084430946959380/642426447167881246/1255641060953161838

Reply:
Yeah, gotcha. There's quite a few of these generic patterns in trivial type conversions (e.g. in strip quantize types or legalize tf types) so I expect that we need one, we'll keep the pattern in utils and reuse it.